tests/framework/static_analysis.py

```diff
@@ -0,0 +1,618 @@
+    ) -> str | int | Tuple[str, Callable[[int], int]]:
+        """
+        If this instruction loads an immediate into the given register, returns
+        that immediate as an integer. If the instruction is a register to register transfer,
+        returns the source register for this transfer. If this instruction doesn't change
+        the given register, returns the given register. Returns None if we don't know
+        how to backpropagate through this instruction.
+
+        :param reg: the register to backpropagate through this instruction
+        :return: An integer if the register is loaded with an immediate by this instruction, or a register
+                 which needs to be backpropagated further (together with an optional forward-propagation
+                 function).
+        """
+
+    def __str__(self):
+        return f"{self.mnemonic} {','.join(self.args)}"
+
+
+class InstructionX86_64(Instruction):  # pylint: disable=invalid-name
+    """A x86_64 instruction"""
+
+    comment_prefix = "#"
+
+    @property
+    def is_call(self):
+        return self.mnemonic in ["call", "jmp"]
+
+    @property
+    def is_syscall(self):
+        return self.mnemonic == "syscall"
+
+    def backpropagate_register(
+        self, reg: str
+    ) -> str | int | Tuple[str, Callable[[int], int]]:
+        # Simplifying assumption: an instruction will not modify a register
+        # that it doesn't reference (generally wrong, but fine for our purposes)
+        affected_registers = [
+            match for (match, _) in re.findall(r"(%[a-z0-9]{2,4})(\W|)", str(self))
+        ]
+        if reg not in affected_registers:
+            return reg
+
+        match self.mnemonic:
+            case "mov":
+                if len(self.args) != 2:
+                    raise UnsupportedInstructionError(self, reg)
+
+                src, dst = self.args
+
+                if dst == reg:
+                    # an immediate load
+                    if src.startswith("$"):
+                        return int(src[3:], 16)
+                    # We moved something into our target register. If it's a new register, we understand
+                    # what's going on. Anything else, and tough luck
+                    if re.match(r"^%\w{2,4}$", src):
+                        return src
+                    raise UnsupportedInstructionError(self, reg)
+                return reg
+            case "xor":
+                src, dst = self.args
+
+                if src == dst:
+                    # we know that reg is part of the arguments, and we know that the arguments are identical
+                    # Thus we have xor reg,reg, which is effectively zeroing reg
+                    return 0
+            case "push":
+                # a push doesn't do anything
+                return reg
+
+        raise UnsupportedInstructionError(self, reg)
+
+
+class InstructionAarch64(Instruction):
+    """An aarch64 assembly instruction"""
+
+    comment_prefix = "//"
```

Comment:
Mh, I actually do not know. Because they are "overwrites" of the field in the abstract base class, so a sort of polymorphic constant (In Rust, `Instruction` would be a trait`, and `comment_prefix` an associated constant). Not sure what the convention here is in python admittedly.

Reply:
It's [looks like](https://stackoverflow.com/questions/58384436/how-to-name-class-variables) there's no strict convention, with some suggestion of lowercase for class variable, and leave all-uppercase for top-level constants.